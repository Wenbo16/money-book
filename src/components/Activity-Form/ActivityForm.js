import React, {useState, useRef, memo} from 'react'
import PropTypes from 'prop-types'
import { isValidDate } from '../../utility'


// edit mode default value based on Item props, so it has to be unconotrolled form
const ActivityForm = memo(({item, onFormSubmit, onCancelSubmit}) => {
    console.log('ActivityForm')
    const[errMessage, setErrMessage] = useState('');

    // useEffect(() => {
    //     setTitle(item.title);
    //     setAmount(item.amount);
    //     setDate(item.date);
    // }, [item])

    const amountInput = useRef(null);
    const dateInput = useRef(null);
    const titleInput = useRef(null);

    const submitForm = (e) => {
        e.preventDefault()
        const editMode = !!item.id
        const amount = amountInput.current.value.trim() * 1
        const date = dateInput.current.value.trim()
        const title = titleInput.current.value.trim()
        
        if (!amount || !date || !title) {
            setErrMessage('请输入所有必选项')
        } else {
            if (amount < 0) {
                setErrMessage('价格数字必须大于0')
            } else if (!isValidDate(date)) {
                setErrMessage('请填写正确的日期格式')
            } else {
                setErrMessage('')
                if (editMode) {
                    onFormSubmit({ ...item, title: title, amount: amount, date: date }, editMode)
                } else {
                    onFormSubmit({ title: title, amount: amount, date: date }, editMode)
                }
            }
        }
       
    }
    const { title, amount, date } = item;
    return (
        <div>
            <form>
                <div className="form-group">
                    <label htmlFor="activity-form-title">Title</label>
                    <input 
                        type="text" 
                        className="form-control" 
                        id="activity-form-title"
                        placeholder="请输入标题"
                        defaultValue={title}
                        ref={titleInput}
                    />
                </div>
                <div className="form-group">
                    <label htmlFor="activity-form-amount">Amount</label>
                    <input 
                        type="number"
                        className="form-control" 
                        id="activity-form-amount" 
                        placeholder="请输入金额" 
                        defaultValue={amount}
                        ref={amountInput}
                    />
                </div>
                <div className="form-group">
                    <label htmlFor="activity-form-date">Date</label>
                    <input 
                        type="date"
                        className="form-control"
                        id="activity-form-date"
                        defaultValue={date}
                        placeholder="请输入日期"
                        ref={dateInput}
                    />
                </div>
                <br/>
                <button type="submit" className="btn btn-primary" onClick={submitForm}>Submit</button>
                <button className="btn btn-secondary" onClick={onCancelSubmit}>Cancel</button>
            </form>
            <br/>
            {errMessage && <div className="alert alert-danger" role="alert">
                {errMessage}
            </div>}
        </div>
    )

})

ActivityForm.propTypes = {
    onFormSubmit: PropTypes.func.isRequired,
    onCancelSubmit: PropTypes.func.isRequired,
    item: PropTypes.object,
}

export default ActivityForm;
