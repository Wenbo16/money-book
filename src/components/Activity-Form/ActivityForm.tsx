import React, { useState, useRef, memo, useEffect } from 'react';
import { isValidDate } from '../../utils/utility';
import { Item } from '../../types';

interface ActivityFormProps {
  onFormSubmit: (data: any, isEdit: boolean) => void;
  onCancelSubmit: () => void;
  item: Item | {};
}

// edit mode default value based on Item props, so it has to be unconotrolled form
const ActivityForm = memo(
  ({ item, onFormSubmit, onCancelSubmit }: ActivityFormProps) => {
    const [errMessage, setErrMessage] = useState('');
    const [amount, setAmount] = useState<number>(0);
    const [date, setDate] = useState<string>('');
    const [title, setTitle] = useState<string>('');

    useEffect(() => {
      setTitle(isItem(item) ? item?.date : '');
      setAmount(isItem(item) ? item?.amount : 0);
      setDate(isItem(item) ? item?.date : '');
    }, [item]);

    const isItem = (value: any): value is Item => value?.id;

    const submitForm = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
      e.preventDefault();
      const editMode = !!isItem(item);
      if (amount === 0 || !date || !title) {
        setErrMessage('请输入所有必选项');
      } else {
        if (amount < 0) {
          setErrMessage('价格数字必须大于0');
        } else if (!isValidDate(date?.trim())) {
          setErrMessage('请填写正确的日期格式');
        } else {
          setErrMessage('');
          if (editMode) {
            onFormSubmit(
              {
                ...item,
                title: title?.trim(),
                amount: amount,
                date: date?.trim(),
              },
              editMode
            );
          } else {
            onFormSubmit(
              { title: title?.trim(), amount: amount, date: date },
              editMode
            );
          }
        }
      }
    };

    return (
      <div>
        <form name="activity-form">
          <div className="form-group">
            <label htmlFor="activity-form-title">Title</label>
            <input
              type="text"
              className="form-control"
              id="activity-form-title"
              name="activity-form-title"
              placeholder="请输入标题"
              onChange={(e) => setTitle(e.target.value)}
              value={title}
            />
          </div>
          <div className="form-group">
            <label htmlFor="activity-form-amount">Amount</label>
            <input
              type="number"
              className="form-control"
              id="activity-form-amount"
              name="activity-form-amount"
              placeholder="请输入金额"
              onChange={(e) => setAmount(Number(e.target.value))}
              value={amount}
            />
          </div>
          <div className="form-group">
            <label htmlFor="activity-form-date">Date</label>
            <input
              type="date"
              className="form-control"
              id="activity-form-date"
              name="activity-form-date"
              placeholder="请输入日期"
              onChange={(e) => setDate(e.target.value)}
              value={date}
            />
          </div>
          <br />
          <button
            type="submit"
            className="btn btn-primary"
            onClick={submitForm}
          >
            Submit
          </button>
          <button className="btn btn-secondary" onClick={onCancelSubmit}>
            Cancel
          </button>
        </form>
        <br />
        {errMessage && (
          <div className="alert alert-danger" role="alert">
            {errMessage}
          </div>
        )}
      </div>
    );
  }
);

export default ActivityForm;
