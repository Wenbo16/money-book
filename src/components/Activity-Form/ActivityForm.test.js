import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
// import RTL from '../../pages/';

import ActivityForm from './ActivityForm';
import { testItems } from '../../testData';
let props = {
  onFormSubmit: jest.fn(),
  onCancelSubmit: jest.fn(),
};
let props_with_item = {
  item: testItems[0],
  onFormSubmit: jest.fn(),
  onCancelSubmit: jest.fn(),
};

export const getInputValue = (selector, wrapper) =>
  wrapper.find(selector).instance().value;

export const setInputValue = (selector, newValue, wrapper) => {
  wrapper.find(selector).instance().value = newValue;
};

describe('test ActivityForm component', () => {
  it('should render the component to match snapshot', () => {
    expect(
      renderer.create(<ActivityForm {...props} />).toJSON()
    ).toMatchSnapshot();
    expect(
      renderer.create(<ActivityForm {...props_with_item} />).toJSON()
    ).toMatchSnapshot();
  });
});

describe('test ActivityForm with no item', () => {
  beforeEach(() => {
    render(<ActivityForm {...props} />);
  });
  it('render ActivityForm should generate three inputs and one form element', () => {
    expect(screen.getByLabelText('Title')).toBeInTheDocument();
    expect(screen.getByLabelText('Amount')).toBeInTheDocument();
    expect(screen.getByLabelText('Date')).toBeInTheDocument();
  });
  it('render ActivityForm with no data should render three inputs and no value', () => {
    expect(screen.getByRole('form')).toHaveFormValues({
      'activity-form-title': '',
      'activity-form-amount': 0,
      'activity-form-date': '',
    });
  });

  it('submit form with empty input should show alert message', () => {
    fireEvent.click(screen.getByText('Submit'));
    expect(screen.getByRole('alert')).toBeVisible();
    expect(props_with_item.onFormSubmit).not.toHaveBeenCalled();
  });

  it('submit form with invalid amount should show alert message', () => {
    render(<ActivityForm {...props} />);
    fireEvent.change(screen.getByLabelText('Title'), {
      target: { value: 'teste' },
    });
    fireEvent.change(screen.getByLabelText('Amount'), {
      target: { value: '-20' },
    });

    // fireEvent.click(screen.getByText('Submit'));
    screen.debug();
    expect(screen.getByRole('form')).toBeInvalid();
  });

  // it('submit form with invalid date format should show alert message', () => {
  //   setInputValue('#title', 'test', wrapper);
  //   setInputValue('#amount', '20', wrapper);
  //   setInputValue('#date', 'wrong date', wrapper);
  //   fireEvent.click(screen.getByText('Submit'));
  //   //   expect(formInstance.state.validatePass).toEqual(false)
  // });
  // it('submit form with valid data should call with the right object', () => {
  //   setInputValue('#title', 'test', wrapper);
  //   setInputValue('#amount', '20', wrapper);
  //   setInputValue('#date', '2018-01-01', wrapper);
  //   const newItem = { title: 'test', amount: 20, date: '2018-01-01' };
  //   wrapper.find('form').simulate('submit');
  //   expect(props.onFormSubmit).toHaveBeenCalledWith(newItem, false);
  // });
  // it('click the cancel button should call the right callback', () => {
  //   wrapper.find('button').last().simulate('click');
  //   expect(props.onCancelSubmit).toHaveBeenCalled();
  // });
  // });
  // describe('test ActivityForm with item data', () => {
  //   it('render ActivityForm with item should render the correct data to inputs', () => {
  //     expect(getInputValue('#title', wrapper2)).toEqual(testItems[0].title);
  //     expect(getInputValue('#amount', wrapper2)).toEqual(
  //       testItems[0].amount.toString()
  //     );
  //     expect(getInputValue('#date', wrapper2)).toEqual(testItems[0].date);
  //   });
  //   it('submit with changed value, should trigger callback with right object', () => {
  //     setInputValue('#title', 'new title', wrapper2);
  //     setInputValue('#amount', '200', wrapper2);
  //     wrapper2.find('form').simulate('submit');
  //     const newItem = { ...testItems[0], title: 'new title', amount: 200 };
  //     //   expect(formInstance.state.validatePass).toEqual(true)
  //     expect(wrapper2.find('.alert').length).toEqual(0);
  //     expect(props_with_item.onFormSubmit).toHaveBeenCalledWith(newItem, true);
  //   });
});
