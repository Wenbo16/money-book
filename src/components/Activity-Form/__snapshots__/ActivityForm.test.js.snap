// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ActivityForm component should render the component to match snapshot 1`] = `
<div>
  <form
    name="activity-form"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="activity-form-title"
      >
        Title
      </label>
      <input
        className="form-control"
        id="activity-form-title"
        name="activity-form-title"
        onChange={[Function]}
        placeholder="请输入标题"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="activity-form-amount"
      >
        Amount
      </label>
      <input
        className="form-control"
        id="activity-form-amount"
        name="activity-form-amount"
        onChange={[Function]}
        placeholder="请输入金额"
        type="number"
        value={0}
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="activity-form-date"
      >
        Date
      </label>
      <input
        className="form-control"
        id="activity-form-date"
        name="activity-form-date"
        onChange={[Function]}
        placeholder="请输入日期"
        type="date"
        value=""
      />
    </div>
    <br />
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
    <button
      className="btn btn-secondary"
      onClick={[MockFunction]}
    >
      Cancel
    </button>
  </form>
  <br />
  
</div>
`;

exports[`test ActivityForm component should render the component to match snapshot 2`] = `
<div>
  <form
    name="activity-form"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="activity-form-title"
      >
        Title
      </label>
      <input
        className="form-control"
        id="activity-form-title"
        name="activity-form-title"
        onChange={[Function]}
        placeholder="请输入标题"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="activity-form-amount"
      >
        Amount
      </label>
      <input
        className="form-control"
        id="activity-form-amount"
        name="activity-form-amount"
        onChange={[Function]}
        placeholder="请输入金额"
        type="number"
        value={0}
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="activity-form-date"
      >
        Date
      </label>
      <input
        className="form-control"
        id="activity-form-date"
        name="activity-form-date"
        onChange={[Function]}
        placeholder="请输入日期"
        type="date"
        value=""
      />
    </div>
    <br />
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
    <button
      className="btn btn-secondary"
      onClick={[MockFunction]}
    >
      Cancel
    </button>
  </form>
  <br />
  
</div>
`;
